Python/compile.c

```diff
@@ -7527,22 +7528,25 @@ _PyCompile_ConstCacheMergeOne(PyObject *const_cache, PyObject **obj)
         return ERROR;
     }
 
-    // t is borrowed reference
-    PyObject *t = PyDict_SetDefault(const_cache, key, key);
+    PyObject *t;
+    int res = PyDict_SetDefaultRef(const_cache, key, key, &t);
     Py_DECREF(key);
-    if (t == NULL) {
+    if (res < 0) {
         return ERROR;
     }
-    if (t == key) {  // obj is new constant.
+    if (res == 0) {  // inserted: obj is new constant.
+        Py_DECREF(t);
         return SUCCESS;
     }
 
     if (PyTuple_CheckExact(t)) {
-        // t is still borrowed reference
-        t = PyTuple_GET_ITEM(t, 1);
+        Py_SETREF(*obj, Py_NewRef(PyTuple_GET_ITEM(t, 1)));
```

Comment:
I'd consider breaking this up, for readability:

```suggestion
        PyObject *item = PyTuple_GET_ITEM(t, 1);
        Py_SETREF(*obj, Py_NewRef(item));
```